crates/ibc/src/services/core/client.rs

```diff
@@ -0,0 +1,314 @@
+use core::str::FromStr;
+use std::boxed::Box;
+
+use ibc_proto::google::protobuf::Any;
+use ibc_proto::ibc::core::client::v1::query_server::Query as ClientQuery;
+use ibc_proto::ibc::core::client::v1::{
+    ConsensusStateWithHeight, IdentifiedClientState, Params as ClientParams,
+    QueryClientParamsRequest, QueryClientParamsResponse, QueryClientStateRequest,
+    QueryClientStateResponse, QueryClientStatesRequest, QueryClientStatesResponse,
+    QueryClientStatusRequest, QueryClientStatusResponse, QueryConsensusStateHeightsRequest,
+    QueryConsensusStateHeightsResponse, QueryConsensusStateRequest, QueryConsensusStateResponse,
+    QueryConsensusStatesRequest, QueryConsensusStatesResponse, QueryUpgradedClientStateRequest,
+    QueryUpgradedClientStateResponse, QueryUpgradedConsensusStateRequest,
+    QueryUpgradedConsensusStateResponse,
+};
+use tonic::{Request, Response, Status};
+use tracing::trace;
+
+use crate::core::ics02_client::error::ClientError;
+use crate::core::ics24_host::identifier::ClientId;
+use crate::core::ics24_host::path::{
+    ClientConsensusStatePath, ClientStatePath, Path, UpgradeClientPath,
+};
+use crate::core::{ContextError, ValidationContext};
+use crate::hosts::tendermint::upgrade_proposal::UpgradeValidationContext;
+use crate::prelude::*;
+use crate::services::core::context::QueryContext;
+use crate::Height;
+
+pub struct ClientQueryService<I, U>
+where
+    I: QueryContext + Send + Sync + 'static,
+    U: UpgradeValidationContext + Send + Sync + 'static,
```

Comment:
Is it necessary for `I` to be bounded by `Send + Sync + static` here and in `ConnectionQuerySerive` and `ChannelQueryService` as well?

Reply:
`Send + Sync + 'static` is needed for the tonic gRPC server trait. It is possible to loosen the bound on the struct generics but it is needed on `Self` when `ClientQuery` is implemented.